src/file_pipe_log/pipe.rs

```diff
@@ -272,31 +370,71 @@ impl<F: FileSystem> SinglePipe<F> {
     }
 
     fn total_size(&self) -> usize {
         let files = self.files.read();
-        (files.active_seq - files.first_seq + 1) as usize * self.target_file_size
+        files.fds.len() * self.target_file_size
     }
 
     fn rotate(&self) -> Result<()> {
         self.rotate_imp(&mut self.active_file.lock())
     }
 
     fn purge_to(&self, file_seq: FileSeq) -> Result<usize> {
-        let (purged, remained) = {
+        let (
+            first_purge_seq, /* first seq for purging */
+            purged,          /* count of purged files */
+            recycled,        /* count of recycled files */
+            remained,        /* count of remained files */
+        ) = {
             let mut files = self.files.write();
-            if file_seq > files.active_seq {
+            if file_seq >= files.first_seq + files.fds.len() as u64 {
+                #[cfg(feature = "failpoints")]
+                {
+                    fail::fail_point!("file_pipe_log::pipe::force_purge", |_| {
+                        let purged = file_seq.saturating_sub(files.first_seq) as usize;
+                        files.fds.drain(..purged);
+                        files.first_seq = file_seq;
+                        files.first_seq_in_use = file_seq;
+                        Ok(purged)
+                    });
+                }
                 return Err(box_err!("Purge active or newer files"));
             }
-            let end_offset = file_seq.saturating_sub(files.first_seq) as usize;
-            files.fds.drain(..end_offset);
-            files.first_seq = file_seq;
-            (end_offset, files.fds.len())
+            println!("Force to do purge op: file_seq: {}", file_seq);
+            let first_purge_seq: u64;
+            let purged: usize;
+            let mut recycled: usize = 0;
+            if files.capacity == 0 {
+                // Not capable for Recycle
+                first_purge_seq = files.first_seq;
+                purged = file_seq.saturating_sub(files.first_seq) as usize;
+                files.fds.drain(..purged);
+                files.first_seq = file_seq;
+                files.first_seq_in_use = file_seq;
+            } else {
+                // Capable for recycling log files. It means that the
+                // FileCollections has a finite volume for storing recycled files.
+                let rest_volume = (files.capacity as u64)
+                    .saturating_sub(files.first_seq_in_use - files.first_seq)
+                    as usize;
+                let expected_purge_count = file_seq.saturating_sub(files.first_seq_in_use) as usize;
+                recycled = std::cmp::min(rest_volume, expected_purge_count);
+                purged = std::cmp::max(rest_volume, expected_purge_count) - rest_volume;
+                first_purge_seq = files.first_seq;
+                files.first_seq += purged as u64;
+                files.first_seq_in_use = file_seq;
+                files.fds.drain(..purged);
+            }
```

Comment:
```suggestion
            if (file_seq <= files.first_seq_in_use) {
                return Ok(0);
            }
            let logically_purged = file_seq - files.first_seq_in_use;
            files.first_seq_in_use = file_seq;
            // Remove some obsolete files if capacity is exceeded.
            let obsolete_files = file.first_seq_in_use - files.first_seq;
            // When capacity is zero, always remove logically deleted files.
            let capacity_exceeded = files.fds.len().saturating_sub(files.capacity);
            let purged = std::cmp::min(capacity_exceeded, obsolete_files);
            (files.first_seq, purged, logically_purged, files.fds.len())
```

Reply:
Not a better choice. I prefer the original one.
With the new algorithm, the `capacity` collided with the original definition of `recycle_capacity`. As the original design shows, the `recycle_capacity` is the restriction on the length of obsolete logs. But here with then new algorithm, it's converted to be the limit of `fds.len()`.

Reply:
I don't understand your argument. Your old implementation is way too complicated. Each `saturating_sub` or `min` or `max` introduces a two-way branching. You have 4 of them (2 `saturating_sub`, 1 `min`, 1 `max`), and 5 branches in total. That comes out 2^5 possibility for me as a reviewer to consider.